### ISSUE TITLE
Integrate into WP EasyCart

### ISSUE THREAD

**Issue by wpeasycart** [affiliatewp/AffiliateWP#557]
Support recommended I leave a note here, we are requesting integration with WP EasyCart.  We have already integrated custom rules for products -> affiliates with special rates, but feel the integration may be warranted and better integrated into Affiliate WP with it's current integration system.

I can gladly get any license or software needed, it can be downloaded via wordpress.org.

**Comment by pippinsplugins** [comment]
Since it is already done in your plugin, would you be willing to post the
code for the integration? That'd be the easiest and fastest way to do it :)

**Comment by wpeasycart** [comment]
We can honestly leave it in our plugin, but my only concern was IF a new affiliate signs up, can they be put into a specific ‘Affiliate Rule’ automatically within your ‘integrations’.   In ours, they are not, we have to add an affiliate to the ‘rule’ to make it work, which could of course be missed or a task if there were a lot of sign-ups.

What I was hoping is on your integration side, customer signs up and is put into the default affiliate rule (product X with Y percentage, product B with X amount, etc.)

If your side integration doesn’t do this either, then we will just leave it as is…  our rules work just like rules, which is kind of funny as we developed it this last month and then looked and saw you did that for other integrations ☺


*********************
Josh Jones
Level Four Development, llc
*********************

**Comment by pippinsplugins** [comment]
We do have certain extensions for controlling the rate that is given to affiliates.

In order to maintain compatibility going forward as changes are made and new features are introduced, it would probably be easiest if the integration existed inside AffiliateWP. Since you wrote it initially, it's obviously entirely up to you, but we would be more than happy to take over the integration. If we do that, we will be responsible for ensuring it stays up to date and continues to work with both platforms.

**Comment by wpeasycart** [comment]
Lets transfer things over so that we can remove the rules option we just put in, better now than later. 

So what I am thinking is if you can give a function to call, say affiliate_wp( )->wpeasycart( $description, $reference, $cart_array ); - we could send in our cart IF your plugin is installed on the user's WordPress install? Right now we call a function at the point that the order is inserted, which uses affiliatewp options + wp easycart rules and processes the cart to add affiliate dollars to affiliatewp if needed. We could instead send the cart to you for processing. Also, if you just let us know what format you want and what data to send, we can send that and let you do the processing. Let me know if this will work or if you have better ideas here.

**Comment by pippinsplugins** [comment]
It'd be best for us to handle it like our other integrations.

For example, you can see how WP e-Commerce works here: https://github.com/affiliatewp/AffiliateWP/blob/master/includes/integrations/class-wpec.php

Does WP Easy Cart have the necessary hooks / filters in place to do it that way?

**Comment by wpeasycart** [comment]
We do not use any hooks or filters, but instead provide a place to call the functions you have here. I'm guessing it would be easy enough to setup a hook to work like you have it setup with the example class?

The one question I do have about the code provided is do you offer line item referrals instead of only the overall order total? The one thing we integrated was a way to provide referrals limited to specific products in the cart. So for example, we can allow referral dollars for our plugin sales, but not custom programming products.

**Comment by pippinsplugins** [comment]
Without hooks, wit will be nearly impossible for us to handle the integration. Personally, I would suggest that you add the necessary hooks so that we can tie into the plugin.

We do support line item referrals, but only for some integrations. Take a look at WooCommerce for example: https://github.com/affiliatewp/AffiliateWP/blob/master/includes/integrations/class-woocommerce.php

**Comment by wpeasycart** [comment]
That looks like just what we need there. I will add the necessary hooks and send those to you later today. It looks like the only thing we will be losing is the ability to create rules that apply to different products and users, but I think that system may be too complicated for most users anyway. Thanks for the info and I'll get back to you soon on this.

**Comment by pippinsplugins** [comment]
Sounds good!

**Comment by wpeasycart** [comment]
We are working on releasing a new version tonight that includes the required hooks. I have also written a class that should be working with the cart. Is there a good place to send the class or would you prefer me to just paste it into this thread?

**Comment by wpeasycart** [comment]
I just wanted to check back in on this, we are ready to move forward with integration and I have a good working example that would fit into AffiliateWP if you would like to see that file? Let us know what we can do to help things move forward.

**Comment by pippinsplugins** [comment]
Sorry fro being slow.

If you could paste it here, that'd be just fine :)

**Comment by wpeasycart** [comment]
<?php

class Affiliate_WP_EasyCart extends Affiliate_WP_Base {

	/**
	 * Setup actions and filters
	 *
	 * @access  public
	 * @since   1.0
	*/
	public function init() {

		$this->context = 'wpeasycart';

		add_action( 'wpeasycart_order_inserted', array( $this, 'add_pending_referral' ), 10, 5 );

		// There should be an option to choose which of these is used
		add_action( 'wpeasycart_order_paid', array( $this, 'mark_referral_complete' ), 10 );
		add_action( 'wpeasycart_full_order_refund', array( $this, 'revoke_referral_on_refund' ), 10 );

	}

	/**
	 * Store a pending referral when a new order is created
	 *
	 * @access  public
	 * @since   1.0
	*/
	public function add_pending_referral(  $order_id, $cart, $order_totals, $user, $payment_type ){

		// Check if an affiliate coupon was used
		$affiliate_id = $this->get_coupon_affiliate_id();

		if( $this->was_referred() || $affiliate_id ) {

			if( false !== $affiliate_id ) {
				$this->affiliate_id = $affiliate_id;
			}

			if( affwp_get_affiliate_email( $this->affiliate_id ) == $user->email ) {
				return false; // Customers cannot refer themselves
			}

			if( affiliate_wp()->referrals->get_by( 'reference', $order_id, $this->context ) ) {
				return false; // Referral already created for this reference
			}

			$cart_shipping = $order_totals->shipping_total;
			$cart_tax = $order_totals->tax_total;

			$items = $cart->cart;

			// Calculate the referral amount based on product prices
			$amount = 0.00;
			foreach( $items as $cart_item ) {

				if( $cart_item->has_affiliate_rule ) {
					continue; // Referrals are disabled on this product
				}

				// The order discount has to be divided across the items

				$product_total = $cart_item->total_price;
				$shipping      = 0;

				if( $cart_shipping > 0 && ! affiliate_wp()->settings->get( 'exclude_shipping' ) ) {

					$shipping       = $cart_shipping / count( $items );
					$product_total += $shipping;

				}

				if( ! affiliate_wp()->settings->get( 'exclude_tax' ) ) {

					$tax       = $cart_tax / count( $items );
					$product_total += $tax;

				}

				if( $product_total <= 0 ) {
					continue;
				}

				$amount += $this->calculate_referral_amount( $product_total, $order_id, $cart_item->product_id );

			}

			if( 0 == $amount && affiliate_wp()->settings->get( 'ignore_zero_referrals' ) ) {
				return false; // Ignore a zero amount referral
			}

			$description = $this->get_referral_description( $cart->cart );
			$visit_id    = affiliate_wp()->tracking->get_visit_id();

			$referral_id = affiliate_wp()->referrals->add( apply_filters( 'affwp_insert_pending_referral', array(
				'amount'       => $amount,
				'reference'    => $order_id,
				'description'  => $description,
				'affiliate_id' => $this->affiliate_id,
				'visit_id'     => $visit_id,
				'context'      => $this->context
			), $amount, $order_id, $description, $this->affiliate_id, $visit_id, array(), $this->context ) );

			if( $referral_id ) {

				$amount = affwp_currency_filter( affwp_format_amount( $amount ) );
				$name   = affiliate_wp()->affiliates->get_affiliate_name( $this->affiliate_id );

				$this->order->add_order_note( sprintf( __( 'Referral #%d for %s recorded for %s', 'affiliate-wp' ), $referral_id, $amount, $name ) );

			}

		}

	}

	/**
	 * Mark referral as complete when payment is completed
	 *
	 * @access  public
	 * @since   1.0
	*/
	public function mark_referral_complete( $order_id = 0 ) {

		$this->complete_referral( $order_id );

	}

	/**
	 * Revoke the referral when the order is refunded
	 *
	 * @access  public
	 * @since   1.0
	*/
	public function revoke_referral_on_refund( $order_id = 0 ) {

		if( ! affiliate_wp()->settings->get( 'revoke_on_refund' ) ) {
			return;
		}

		if( 'shop_order' != get_post_type( $order_id ) ) {
			return;
		}

		$this->reject_referral( $order_id );

	}

	/**
	 * Retrieves the referral description
	 *
	 * @access  public
	 * @since   1.1
	*/
	public function get_referral_description( $items ) {

		$description = array();

		foreach ( $items as $key => $item ) {

			$description[] = $item->title;

		}

		$description = implode( ', ', $description );

		return $description;

	}

}
new Affiliate_WP_EasyCart;

**Comment by pippinsplugins** [comment]
Excellent, that'll do!

**Comment by pippinsplugins** [comment]
@wpeasycart Which status fires the `wpeasycart_order_paid` hook? https://cloudup.com/cAoejX4QOXh

**Comment by pippinsplugins** [comment]
The `revoke_referral_on_refund()` doesn't seem to be firing. Can you look at it and confirm it's setup properly?

**Comment by wpeasycart** [comment]
It looks like you are using the free version, which we didn't implement this feature with. I think we will add it to the free version this week though to prevent confusion. 

To really test this you should install our admin plugin, which is referenced in a notice at the top of the screen or go to easycart admin -> store admin and download from there. I set up a key for you to use pippensplugins to register your site as a full version. 

The order statuses for wpeasycart_order_paid fire on card approved and direct deposit received (we are adding that to order confirmed and third party approved as well). 

The revoke on refund will only occur on a full refund. 

Let me know if you have questions

**Comment by pippinsplugins** [comment]
I installed the free admin plugin.

Can you let me know when you add the hooks to the free version?

**Comment by pippinsplugins** [comment]
@wpeasycart Do you think this will be ready in the next couple of days? We're releasing 1.6 on Wednesday, so if it's not ready by then, we'll need to punt this to 1.7.

**Comment by pippinsplugins** [comment]
We're going to push this back to 1.6.1/2/3 since it's not quite there yet and 1.6 needs to go out tomorrow.

**Comment by wpeasycart** [comment]
Bad timing on this, we just released our update last night No worries though, whenever you get a chance to implement and include in a release will be fine. It should hook with the free version, lite, and standard. Let us know how it goes with the free version.

**Comment by pippinsplugins** [comment]
Oh sweet, we'll include this in 1.6.1 then!

**Comment by wpeasycart** [comment]
Hello, just wanted to check in and see how things are coming with your integration with WP EasyCart? Any estimate on a release date that will include this? Thank you!

**Comment by pippinsplugins** [comment]
We're pushing 1.6.2 in a couple of days and it should be included with that.

**Comment by wpeasycart** [comment]
Great! Thank you, we will get ready to push some information on it.

**Comment by wpeasycart** [comment]
Could we get an update on this project, seems maybe this has been moved to 1.6.3? If we can help move this along, please let us know.

**Comment by pippinsplugins** [comment]
Sorry for the delay, I'm looking at this again now.

**Comment by pippinsplugins** [comment]
@wpeasycart You mentioned that you set up a key for me. Is `pippensplugins` the domain or user? Can you tell me the user / password to use?

**Comment by wpeasycart** [comment]
this is a license key that you can use on any domain to unlock the full version of the software. Register this by installing our store admin software (explained in the plugin when you go to the easycart admin -> store admin page) and register with the key.

**Comment by pippinsplugins** [comment]
@wpeasycart It seems the latest version (from WordPress.org) has a fatal error, at least with multisite.

I get this as soon as I activate it: Fatal error: Class 'ec_categorywidget' not found in /Users/pippin/Sites/edd/wp-includes/widgets.php on line 560

**Comment by pippinsplugins** [comment]
We're pushing 1.6.3 out tonight. I'm sorry but without a functional version of WP EasyCart, we cannot do final testing to put this in. Punting to 1.7.

**Comment by wpeasycart** [comment]
Pippin, never saw these last few messages... ya, easycart is a single site system, not multi-site compatible at this time...  Do you got any single site wordpress test sites you can work on and register easycart with, or do you want us to setup a test site for you?  

I know we are getting more and more customers starting to get affiliate wp AND easycart running, hopfully by 1.7 we can put it into your plugin...  right now we have our own rule system running for users.

**Comment by pippinsplugins** [comment]
Why does it fatal error on a single site? I'm not network activating it, just activating on a single site.

**Comment by pippinsplugins** [comment]
@wpeasycart I am working on integrating this further and noticed that you still have your AffiliateWP integration code in place. Are you going to remove it once we push this update out?

**Comment by pippinsplugins** [comment]
@sumobi this is ready for testing.

@wpeasycart we'll need to have the direct integration that you guys have removed as soon as 1.7 goes out near the end of August. Could you take care of ensuring that happens?

**Comment by sumobi** [comment]
@pippinsplugins 

Seems to generate 2 referrals per order:

![screen shot 2015-08-27 at 7 43 14 pm](https://cloud.githubusercontent.com/assets/52581/9515226/19bc4a80-4cf4-11e5-957e-b455225dabd3.png)

___

When I completed the order, the first referral is updated fine, but the other is still there and `pending`:

![screen shot 2015-08-27 at 7 44 03 pm](https://cloud.githubusercontent.com/assets/52581/9515225/199cf1d0-4cf4-11e5-9a39-902f72b36505.png)

___

Just to make sure I made one more order and 2 more were created:

![screen shot 2015-08-27 at 7 44 39 pm](https://cloud.githubusercontent.com/assets/52581/9515224/199b0898-4cf4-11e5-8500-71980b041944.png)

**Comment by pippinsplugins** [comment]
@sumobi that's because of the AffiliateWP integration included in WP EasyCart, hence my comments above about that having to be removed before we can push this out.

**Comment by wpeasycart** [comment]
@pippinsplugins, we can pull the code that is causing the duplicates here at the end of the week if you guys are planning on launching this integrated version? Let us know if you have a good estimate on the date and we will try and match up an update with your release.

**Comment by pippinsplugins** [comment]
Awesome, thanks!

We are hopping to launch 1.7 (with the integration within the next two
weeks).

**Comment by pippinsplugins** [comment]
@wpeasycart let me know when it's been removed and then we can get this merged in.

**Comment by wpeasycart** [comment]
@pippinsplugins we will plan to release an upgrade with this removed on Monday, 8/30 in version 8.1.15.

**Comment by pippinsplugins** [comment]
:+1:

**Comment by pippinsplugins** [comment]
@wpeasycart is the update going out soon? I don't see anything on the .org changelog.

**Comment by wpeasycart** [comment]
Hi @pippinsplugins, sorry for the delay here, pushing a new version in the morning with this change. Need to make a last correction to our admin tools to ensure there is no confusion for our customers.

**Comment by wpeasycart** [comment]
@pippinsplugins just released 3.1.15, which has the AffiliateWP code removed

**Comment by pippinsplugins** [comment]
Great, thanks!

**Comment by pippinsplugins** [comment]
Merged into `release/1.7`
